pkg/api/datasources.go

```diff
@@ -598,19 +600,27 @@ func getEncodedString(jsonData *simplejson.Json, key string) string {
 	return string(val)
 }
 
-func checkTeamHTTPHeaderPermissions(hs *HTTPServer, c *contextmodel.ReqContext, ds *datasources.DataSource, cmd datasources.UpdateDataSourceCommand) (bool, error) {
+func checkTeamHTTPHeadersDiff(ds *datasources.DataSource, jsonData *simplejson.Json) bool {
+	if ds == nil || jsonData == nil {
+		return false
+	}
 	currentTeamHTTPHeaders := getEncodedString(ds.JsonData, "teamHttpHeaders")
-	newTeamHTTPHeaders := getEncodedString(cmd.JsonData, "teamHttpHeaders")
-	if (currentTeamHTTPHeaders != "" || newTeamHTTPHeaders != "") && currentTeamHTTPHeaders != newTeamHTTPHeaders {
-		return evaluateTeamHTTPHeaderPermissions(hs, c, datasources.ScopePrefix+ds.UID)
+	newTeamHTTPHeaders := getEncodedString(jsonData, "teamHttpHeaders")
+	if currentTeamHTTPHeaders == "" && newTeamHTTPHeaders == "" {
+		return false
 	}
-	return true, nil
+	return currentTeamHTTPHeaders != newTeamHTTPHeaders
 }
 
 func (hs *HTTPServer) updateDataSourceByID(c *contextmodel.ReqContext, ds *datasources.DataSource, cmd datasources.UpdateDataSourceCommand) response.Response {
 	if ds.ReadOnly {
 		return response.Error(http.StatusForbidden, "Cannot update read-only data source", nil)
 	}
+	// could get expensive
+	// we only want users to be able to update team HTTP headers from the updateDatasourceLBACRules
+	if checkTeamHTTPHeadersDiff(ds, cmd.JsonData) {
+		return response.Error(http.StatusForbidden, "Cannot update team HTTP headers for data source, need to use updateDatasourceLBACRules API", nil)
+	}
```

Comment:
You already check this in `UpdateDataSourceByID` and `UpdateDataSourceByUID`. There's no need to run the check twice, so perhaps you can just leave this one and remove the other two checks?

Reply:
this is true!

Reply:
we have now removed the diff instead, it made the whole implementation confusing